File: server_mode/client/post.py
import os
import time
import argparse
import pyperclip
import requests


from dotenv import load_dotenv

load_dotenv()


# If content is None, it will read from clipboard and send to server.
def post_content(url, content=None):
    if content is None:
        content = pyperclip.paste()

    content_replaced = (
        content.replace("\n", "<LF>")
        .replace("\r", "<CR>")
        .replace("\t", "<TAB>")
        .replace(" ", "<SPACE>")
    )
    print(f"Send content: `{content_replaced}`")

    print(f"Sending POST request to `{url}`.")
    timeout_seconds = float(os.getenv("REQUEST_TIMEOUT_SECONDS", "5"))

    # Get id
    script_dir = os.path.dirname(os.path.abspath(__file__))
    id_path = os.path.join(script_dir, "id")
    try:
        with open(id_path, "r", encoding="utf-8") as f:
            id = f.read().strip()
    except FileNotFoundError:
        id = ""

    timestamp = str(int(time.time()))
    try:
        response = requests.post(
            url,
            json={"id": id, "text": content_replaced, "timestamp": timestamp},
            timeout=timeout_seconds,
        )
    except requests.exceptions.ConnectionError:
        print("Error: Cannot connect to server.")
        return False
    except requests.exceptions.Timeout:
        print(f"Error: Request timed out after {timeout_seconds} seconds.")
        return False
    except requests.RequestException as error:
        print(f"Error: Failed to send content: {error}")
        return False

    print(f"Server response: `{response.text}`")

    if response.status_code == 200:
        print(f"Clipboard content sent successfully.")
        return True
    else:
        print(f"Failed to send clipboard content. Status code: {response.status_code}")
        return False


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Send text to tcopy server")
    parser.add_argument(
        "content",
        nargs="*",
        help="Text content to send. If omitted, clipboard content is used.",
    )
    args = parser.parse_args()

    server_base_url = os.getenv("SERVER_BASE_URL")

    if server_base_url is None:
        print("Error: SERVER_BASE_URL not found in .env file")
    else:
        content = " ".join(args.content).strip() if args.content else None
        success = post_content(server_base_url, content)
        if not success:
            raise SystemExit(1)

File: server_mode/client/post.js
// Send text (or current clipboard content) to the tcopy server.
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import 'dotenv/config';
import { readId } from '../utils/idUtils.js';
import { readClipboard } from '../utils/clipboardUtils.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const REQUEST_TIMEOUT_MS = parseFloat(process.env.REQUEST_TIMEOUT_SECONDS ?? '5') * 1000;

const args = process.argv.slice(2);
const baseUrl = process.env.SERVER_BASE_URL;

if (!baseUrl) {
  console.error('Error: SERVER_BASE_URL not found in .env file');
  process.exit(1);
}

async function postContent(url, content) {
  const contentReplaced = content
    .replace(/\n/g, '<LF>')
    .replace(/\r/g, '<CR>')
    .replace(/\t/g, '<TAB>')
    .replace(/ /g, '<SPACE>');

  console.log(`Send content: \`${contentReplaced}\``);
  console.log(`Sending POST request to \`${url}\`.`);

  // Read client id
  const id = readId(path.join(__dirname, 'id'));
  const timestamp = String(Math.floor(Date.now() / 1000));

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ id, text: contentReplaced, timestamp }),
      signal: controller.signal,
    });
    clearTimeout(timer);

    const body = await response.text();
    console.log(`Server response: \`${body}\``);

    if (response.ok) {
      console.log('Clipboard content sent.');
      return true;
    } else {
      console.log(`Failed to send clipboard content. Status code: ${response.status}`);
      return false;
    }
  } catch (e) {
    clearTimeout(timer);
    if (e.name === 'AbortError') {
      console.error(`Error: Request timed out after ${REQUEST_TIMEOUT_MS / 1000} seconds.`);
    } else {
      console.error('Error: Cannot connect to server.');
    }
    return false;
  }
}

const content = args.length > 0 ? args.join(' ').trim() : readClipboard();
const success = await postContent(baseUrl, content);
if (!success) process.exit(1);
